src/measure_theory/integral/layercake.lean

```diff
@@ -0,0 +1,260 @@
+  simp_rw [integrand_eq, ← lintegral_indicator (λ t, ennreal.of_real (g t)) measurable_set_Ioc,
+           ← lintegral_indicator _ measurable_set_Ioi],
+  rw lintegral_lintegral_swap,
+  { apply congr_arg,
+    funext s,
+    have aux₁ : (λ x, (Ioc 0 (f x)).indicator (λ (t : ℝ), ennreal.of_real (g t)) s)
+                = (λ x, (ennreal.of_real (g s) * (Ioi (0 : ℝ)).indicator (λ _, 1) s)
+                             * (Ici s).indicator (λ (t : ℝ), (1 : ℝ≥0∞)) (f x)),
+    { funext a,
+      by_cases s ∈ Ioc (0 : ℝ) (f a),
+      { simp only [h, (show s ∈ Ioi (0 : ℝ), from h.1),
+                   (show f a ∈ Ici s, from h.2), indicator_of_mem, mul_one], },
+      { have h_copy := h,
+        simp only [mem_Ioc, not_and, not_le] at h,
+        by_cases h' : 0 < s,
+        { simp only [h_copy, h h', indicator_of_not_mem, not_false_iff, mem_Ici, not_le,
+                     mul_zero], },
+        { have : s ∉ Ioi (0 : ℝ) := h',
+          simp only [this, h', indicator_of_not_mem, not_false_iff, mul_zero, zero_mul, mem_Ioc,
+                     false_and], }, }, },
+    simp_rw aux₁,
+    rw lintegral_const_mul',
+    swap, { apply ennreal.mul_ne_top ennreal.of_real_ne_top,
+            by_cases s ∈ Ioi (0 : ℝ); { simp [h], }, },
+    simp_rw [(show (λ a, (Ici s).indicator (λ (t : ℝ), (1 : ℝ≥0∞)) (f a))
+                   = (λ a, {a : α | s ≤ f a}.indicator (λ _, 1) a),
+              by { funext a, by_cases s ≤ f a; simp [h], })],
+    rw lintegral_indicator,
+    swap, { exact f_mble measurable_set_Ici, },
+    rw [lintegral_one, measure.restrict_apply measurable_set.univ, univ_inter, indicator_mul_left,
+        mul_assoc,
+        (show (Ioi 0).indicator (λ (_x : ℝ), (1 : ℝ≥0∞)) s * μ {a : α | s ≤ f a}
+              = (Ioi 0).indicator (λ (_x : ℝ), 1 * μ {a : α | s ≤ f a}) s,
+        by { by_cases 0 < s; simp [h], })],
+    simp_rw [mul_comm _ (ennreal.of_real _), one_mul],
+    refl, },
+  have aux₂ : function.uncurry
+              (λ (x : α) (y : ℝ), (Ioc 0 (f x)).indicator (λ (t : ℝ), ennreal.of_real (g t)) y)
+              = {p : α × ℝ | p.2 ∈ Ioc 0 (f p.1)}.indicator (λ p, ennreal.of_real (g p.2)),
+  { funext p,
+    cases p,
+    dsimp,
+    by_cases p_snd ∈ Ioc 0 (f p_fst),
+    { have h' : (p_fst, p_snd) ∈ {p : α × ℝ | p.snd ∈ Ioc 0 (f p.fst)} := h,
+      rw [set.indicator_of_mem h', set.indicator_of_mem h], },
+    { have h' : (p_fst, p_snd) ∉ {p : α × ℝ | p.snd ∈ Ioc 0 (f p.fst)} := h,
+      rw [set.indicator_of_not_mem h', set.indicator_of_not_mem h], }, },
+  rw aux₂,
+  have mble := measurable_set_region_between_oc measurable_zero f_mble measurable_set.univ,
+  simp_rw [mem_univ, pi.zero_apply, true_and] at mble,
+  exact (ennreal.measurable_of_real.comp (g_mble.comp measurable_snd)).ae_measurable.indicator mble,
+end
+
+lemma ae_measurable_Ioi_of_forall_Ioc
+  (g_meas : ∀ t > 0, ae_measurable g (volume.restrict (Ioc 0 t))) :
+  ae_measurable g (volume.restrict (Ioi 0)) :=
```

Comment:
We need to move this and the lemma below it to other files. I'll look for good places.

Reply:
I generalized and moved those two lemmas.